Fix(postgres): fallback to parameter parser if heredoc is untokenizable

georgesittas opened the pull request:
Fixes #2931

Postgres requires heredoc tags to follow the same lexical rules as the unquoted identifiers, so a digit after the `HEREDOC_STRING` token (`$`) can unambiguously be assumed to be a parameter number.

Reference: https://www.postgresql.org/docs/current/sql-syntax-lexical.html

izeigerman commented on sqlglot/tokens.py:
```diff
@@ -517,6 +518,9 @@ def _quotes_to_format(
                 string=_TOKEN_TYPE_TO_INDEX[TokenType.STRING],
                 var=_TOKEN_TYPE_TO_INDEX[TokenType.VAR],
+                heredoc_string_alternative=_TOKEN_TYPE_TO_INDEX[
+                    klass.HEREDOC_STRING_ALTERNATIVE or TokenType.VAR
```
```
or TokenType.VAR
```
I don't see the same logic being applied to the Python tokenizer. Is this expected?

izeigerman replied:
Mirroring would be easier to maintain for sure. Less cognitive load

tobymao replied:
agree rust should try to mimic python implementation whenever possible

izeigerman commented on sqlglotrs/src/tokenizer.rs:
```diff
@@ -399,6 +399,19 @@ impl<'a> TokenizerState<'a> {
                 (Some(2), *token_type, end.clone())
             } else if *token_type == self.token_types.heredoc_string {
+                if self.settings.heredoc_tag_is_identifier
+                    && !(self.peek_char.is_alphabetic() || self.peek_char == '_')
```
```
|| self.peek_char == '_'
```
this is not in the Python version either.

georgesittas replied:
That's right, the Python version uses `isidentifier()` which aims to replicate this logic. I searched tokenizer.rs and it seems like the other `isidentifier()` call is mirrored like what I did here, inside `scan_number()`. Should we factor this out in a helper with a similar name? Is there no rust equivalent for alpha + `_` checking?

izeigerman replied:
> Should we factor this out in a helper with a similar name? Is there no rust equivalent for alpha + _ checking?

ah, makes sense. No, I didn't find one

georgesittas commented:
Addressed feedback:

- Python side now mimics Rust side w.r.t. the `HEREDOC_STRING_ALTERNATIVE` logic
- Factored out the `is_alphabetic() ||  ... == '_'` logic so it resembles the Python side more.
